File: internal/parity/fixture.go
package parity

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"os"
)

const maxJSONLLineSize = 64 * 1024 * 1024

// Fixture is one differential test case from a JSONL fixture corpus.
type Fixture struct {
	ID        string          `json:"id"`
	Subsystem string          `json:"subsystem"`
	Input     json.RawMessage `json:"input"`
	Expected  json.RawMessage `json:"expected"`
}

// LoadJSONL reads newline-delimited fixtures from r.
// Blank lines are ignored, while parse errors report their one-based line number.
func LoadJSONL(r io.Reader) ([]Fixture, error) {
	scanner := bufio.NewScanner(r)
	scanner.Buffer(make([]byte, 64*1024), maxJSONLLineSize)

	var fixtures []Fixture
	lineNumber := 0
	for scanner.Scan() {
		lineNumber++
		line := bytes.TrimSpace(scanner.Bytes())
		if len(line) == 0 {
			continue
		}

		var fixture Fixture
		if err := json.Unmarshal(line, &fixture); err != nil {
			return nil, fmt.Errorf("decode JSONL line %d: %w", lineNumber, err)
		}
		fixtures = append(fixtures, fixture)
	}
	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("scan JSONL line %d: %w", lineNumber+1, err)
	}

	return fixtures, nil
}

// LoadFile loads newline-delimited fixtures from path.
func LoadFile(path string) ([]Fixture, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open fixture file %q: %w", path, err)
	}
	defer file.Close()

	fixtures, err := LoadJSONL(file)
	if err != nil {
		return nil, fmt.Errorf("load fixture file %q: %w", path, err)
	}
	return fixtures, nil
}

File: bitmagnet-rs/crates/bitmagnet-diff/src/fixture.rs
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::path::Path;

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use serde_json::Value;

/// One differential test case from a JSONL fixture corpus.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Fixture {
    pub id: String,
    pub subsystem: String,
    pub input: Value,
    pub expected: Value,
}

/// Load newline-delimited fixtures from a buffered reader.
///
/// Blank lines are ignored, while read and parse errors include their
/// one-based line number.
pub fn load_jsonl<R: BufRead>(reader: R) -> Result<Vec<Fixture>> {
    let mut fixtures = Vec::new();

    for (index, line) in reader.lines().enumerate() {
        let line_number = index + 1;
        let line = line.with_context(|| format!("read JSONL line {line_number}"))?;
        if line.trim().is_empty() {
            continue;
        }

        let fixture = serde_json::from_str(&line)
            .with_context(|| format!("parse JSONL line {line_number}"))?;
        fixtures.push(fixture);
    }

    Ok(fixtures)
}

/// Load newline-delimited fixtures from a file path.
pub fn load_file<P: AsRef<Path>>(path: P) -> Result<Vec<Fixture>> {
    let path = path.as_ref();
    let file = File::open(path).with_context(|| format!("open fixture file {}", path.display()))?;

    load_jsonl(BufReader::new(file))
        .with_context(|| format!("load fixture file {}", path.display()))
}
